### DISCUSSION TITLE
Manageable redirects in front-matter

### DISCUSSION THREAD

**Discussion by lilith** [withastro/roadmap#627] [Proposal]
### Body

# Summary

Allow content pages to list inbound redirects in front-matter, like `Aliases: /historical-path`.

# Background & Motivation

File-based routing is wonderful, but redirect management in an external file is not good for long-term maintenance. Even greenfield sites accumulate redirects quickly, and it's best if those mappings have less data that can go stale. Specifically, put the 'from' side in the content frontmatter itself, which ensures the destination is always valid. 

Personally, I have hundreds of pages of content that already uses this approach in a [different CMS with a similar philosophy](https://github.com/imazen/hardwired) and have found it excellent over the last decade. 

Presumable some kind of index and cache will be needed in certain modes, but as a data structure it's just a map. 

# Goals

- Enable a good long-term maintenance story for URLs and redirects in SSR and SSG modes
- Allow migrations of content to Astro that already use this model. 
- Optionally, add a `Redirect-to` frontmatter instruction that goes the opposite way. This would add symmetry to the abstraction, and allow adapters to do the best-available approach (server-side redirects are best when supported by the static host platform, falling back to meta tags when required for less advanced static site servers)
- Optionally, support both temporary and permanent redirects for managing SEO worth rankings. 

Other mentions of this issue
https://github.com/withastro/roadmap/issues/466
https://github.com/withastro/roadmap/issues/466#issuecomment-1435824862
https://github.com/withastro/roadmap/issues/466#issuecomment-1555138194

# Example

```
---
aliases: /legacy-url-1 /legacy+url+2
---
```
or
```
---
permanent-redirect-from: ['/legacy-url-1', '/legacy url 2']
temporary-redirect-from: '/legacy-url-3'
---
```

and 
```
---
temporary-redirect-to: /replacement-for-this-page
---
```

I didn't find multi-word examples of frontmatter, so naming here could use guidance. I currently use space-delimited URLs and "Aliases" in my legacy sites.

**Comment by davidvkimball** [comment]
+1

**Comment by florian-lefebvre** [comment]
I think there's no need to make changes in core for this. You can write your own redirection logic in the astro page that manages this collection.

Do you think there are any blockers in core?

**Comment by davidvkimball** [comment]
@lilith hey, I've actually done this myself with this Astro theme I'm working on, and it works perfectly. 

https://astro-modular.netlify.app/ 

I'm getting pretty close to releasing V1.
